src/blocks/transformers/modalForm/ModalForm.tsx

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2021 PixieBrix, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+import Form from "@rjsf/core";
+import { Message } from "@/core";
+import { browser } from "webextension-polyfill-ts";
+import { useAsyncState } from "@/hooks/common";
+import { whoAmI } from "@/background/messenger/api";
+import {
+  FORM_CANCEL,
+  FORM_GET_DEFINITION,
+  FORM_RESOLVE,
+  FormDefinition,
+} from "@/blocks/transformers/modalForm/formTypes";
+import GridLoader from "react-spinners/GridLoader";
+import { getErrorMessage } from "@/errors";
+
+const ModalLayout: React.FC = ({ children }) => (
+  // Don;t use React Bootstrap's Modal because we want to customize the classes in the layout
+  <div className="modal-content">
+    <div className="modal-body">{children}</div>
+  </div>
+);
+
+const ModalForm: React.FC = () => {
+  const params = new URLSearchParams(location.search);
+  const nonce = params.get("nonce");
+  const sourceFrameId = Number.parseInt(params.get("frameId") ?? "0", 10);
```

Comment:
Assuming that the parameter is a regular, stringified number (as opposed to a string like `"1px"`), you can use `Number()` instead, which has the nice side effect of turning `undefined` into 0, so `?? 0` also isn't needed
```suggestion
  const sourceFrameId = Number(params.get("frameId"));
```

Reply:
> or if it could be replaced by the existing "popup" component

That a good thought. IIRC another limitation is it might not work properly on Firefox?

I'll consider that option as we think through the draggable use case

Reply:
> That a good thought. IIRC another limitation is it might not work properly on Firefox?

Only when in macOS fullscreen, yes